src/GridFS/Exception/StreamException.php

```diff
@@ -0,0 +1,40 @@
+<?php
+
+namespace MongoDB\GridFS\Exception;
+
+use MongoDB\Exception\RuntimeException;
+use function MongoDB\BSON\fromPHP;
+use function MongoDB\BSON\toJSON;
+use function sprintf;
+use function stream_get_meta_data;
+
+class StreamException extends RuntimeException
+{
+    /** @param resource $destination */
+    public static function downloadFromFilenameFailed(string $filename, $destination) : self
+    {
+        $metadata = stream_get_meta_data($destination);
+
+        return new static(sprintf('Downloading GridFS file "%s" to resource "%s" failed: stream error.', $filename, $metadata['uri']));
+    }
+
+    /**
+     * @param mixed    $id
+     * @param resource $destination
+     */
+    public static function downloadFromIdFailed($id, $destination) : self
+    {
+        $stringId = toJSON(fromPHP(['_id' => $id]));
+        $metadata = stream_get_meta_data($destination);
+
+        return new static(sprintf('Downloading GridFS file with identifier "%s" to resource "%s" failed: stream error.', $stringId, $metadata['uri']));
+    }
+
+    /** @param resource $source */
+    public static function uploadFailed(string $filename, $source) : self
```

Comment:
In the case of `openUploadStream`, the path created by `createPathForUpload` only points to the bucket namespace since the filename and ID aren't assembled into a file document until later in the WritableStream constructor; however, by the point you'd throw this exception, that underlying stream would already have been created. If you wanted to provide a path in the same form as a downloaded file exception, we _could_ access the raw file document for the resource with `Bucket::getRawFileDocumentForStream()` and use `createPathForFile` to assemble a URI.

That said, the user knowingly provided a filename string so it's probably best to still include that in the error message, too. Perhaps:

    Uploading file from "<source-uri>" to "<gridfs-uri>" failed. GridFS filename: <filename>

The GridFS URI will always have a "gridfs://" prefix, so it's clear that the destination is a GridFS file (although they technically could both be).

Reply:
Updated to use the error message above 👍